### DISCUSSION TITLE
Parsing SVGs to internal vector format

### DISCUSSION THREAD

**Discussion by i-make-robots** [css4j/echosvg#43] [General]
Hello!  Glad to find a java svg parse that is active.

We're using Batik in [makelangelo software](https://github.com/MarginallyClever/Makelangelo-software) and either it has issues or I'm really bad at reading the documentation.  I'm looking for an alternative that avoids the [issues we've run across](https://github.com/MarginallyClever/Makelangelo-software/issues?q=is%3Aopen+is%3Aissue+label%3ASVG).

Have I come to the right place?

**Comment by carlosame** [comment]
This isn't really a formal bug report, so I'm moving it to Discussions.

**Comment by carlosame** [comment]
> Hello! Glad to find a java svg parse that is active.
> 
> We're using Batik in [makelangelo software](https://github.com/MarginallyClever/Makelangelo-software) and either it has issues or I'm really bad at reading the documentation.

Batik has enough bugs, both in the main code and its testing infrastructure, that one cannot reliably run their test suite. Their tests aren't reproducible, if you do not believe this please try to run their "regard test suite".

IMO that's the main reason why they do not fix bugs unless they have a CVE attached.
 
> Have I come to the right place?

After a lot of work, I managed to transform Batik's tests into something that can be run consistently from CI (at least on Windows). If I write a patch I can reliably check for regressions, so I believe that this project is at least a better place.

I'll look at your issues and see what I can do.

**Comment by carlosame** [comment]
> I'll look at your issues and see what I can do.

Hello Dan,

I've looked at those issues and there seems to be the assumption that Batik is causing certain problems, but I do not see any Batik-specific investigation that I could use. I'm not familiar with your software and do not know what happens under the hood.

I'd suggest that you open one or more issues here, each with a sample SVG image that allows to clearly reproduce the problem (with authorisation to include the image as a new rendering test in this project), and also the expected/obtained rendering. You may want to look at [`SamplesRenderingTest.java`](https://github.com/css4j/echosvg/blob/master/echosvg-test/src/test/java/io/sf/carte/echosvg/test/svg/SamplesRenderingTest.java) for examples of rendering tests; adding a test entry there should be all that you need to trigger a rendering (the tested SVG files are located under `samples`).

The file [`IMAGE_COMPARISONS`](https://github.com/css4j/echosvg/blob/master/IMAGE_COMPARISONS.md) is also useful to deal with the generated image(s). Please ask for help if you need it.

If you aren't rendering but instead expect a specific value being obtained from some EchoSVG/Batik method, you should also give the relevant values and information to reproduce.

I look forward your issues being opened.

**Comment by i-make-robots** [comment]
Thanks for taking a look.  Do I understand correctly that your fork should be a direct replacement for batik?  after changing include paths and module-info, of course.  API the same, implementation changed.

Our project is maven-based.  I see your getting started instructions are all for gradle and build-it-yourself.  is there a maven central dependency I can use?

**Reply by carlosame** [reply]
> Thanks for taking a look. Do I understand correctly that your fork should be a direct replacement for batik? after changing include paths and module-info, of course. API the same, implementation changed.

Yes it is, except for a small set of changes. See [MIGRATING_FROM_BATIK](https://github.com/css4j/echosvg/blob/master/MIGRATING_FROM_BATIK.md).


> Our project is maven-based. I see your getting started instructions are all for gradle and build-it-yourself. is there a maven central dependency I can use?

[None of my projects are on Maven Central](https://groups.google.com/g/css4j/c/op5jIoINb3M/m/IiiN-LfkDAAJ), but you could set up the following repository in your `pom.xml`:

```xml
  <repositories>
    <repository>
        <id>css4j</id>
        <name>CSS4J repository</name>
        <url>https://css4j.github.io/maven/</url>
    </repository>
  </repositories>
```

Although I recommend using Gradle because with Maven you cannot configure which artifacts are downloaded from a given repository.

**Reply by i-make-robots** [reply]
Ok, I'm trying it now.

**Comment by i-make-robots** [comment]
for maven i added
```xml
	<repositories>
		<!-- for echoSVG -->
		<repository>
			<id>css4j</id>
			<name>CSS4J repository</name>
			<url>https://css4j.github.io/maven/</url>
		</repository>
	</repositories>
```
and in module-info.java
```java
    requires io.sf.carte.echosvg;
```
then fixed all import commands.  The only issue stopping my build now is 
```java
        SVGDocument newDocumentFromInputStream(InputStream in) throws Exception {
		String parser = XMLResourceDescriptor.getXMLParserClassName();
		SAXSVGDocumentFactory factory = new SAXSVGDocumentFactory(parser);
		return (SVGDocument) factory.createDocument("",in);
	}
```
The system claims it doesn't recognize `org.apache.batik.util.XMLResourceDescriptor`.  I tried adding
```java
    requires io.sf.carte.echosvg.util;
```
and now i need a hint, please.

**Reply by carlosame** [reply]
> The system claims it doesn't recognize `org.apache.batik.util.XMLResourceDescriptor`

Yes the `String` arguments to the document factories were replaced with a `XMLReader` due to #24. This is consistent with what other Java DOM implementations do.

**Comment by i-make-robots** [comment]
So based on the `migrating from BATIK` doc I tried
```java
private static SVGDocument newDocumentFromInputStream(InputStream in) throws Exception {
	SAXDocumentFactory factory = new SAXSVGDocumentFactory();
	return (SVGDocument)factory.createDocument(null,in);
}
```
and now I get a runtime boot layer failure becaues of `xalan`.

**Reply by carlosame** [reply]
> So based on the `migrating from BATIK` doc I tried
> 
> ```java
> private static SVGDocument newDocumentFromInputStream(InputStream in) throws Exception {
> 	SAXDocumentFactory factory = new SAXSVGDocumentFactory();
> 	return (SVGDocument)factory.createDocument(null,in);
> }
> ```
> 
> and now I get a runtime boot layer failure becaues of `xalan`.

Details please? Is it the known issue due to `xml-apis` being carried by Xalan?

If that is the case, use a [Maven dependency exclusion](https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html) like:

```xml
      <exclusions>
        <exclusion>
          <groupId>xml-apis</groupId>
          <artifactId>xml-apis</artifactId>
        </exclusion>
      </exclusions>
```

Otherwise I'll need more information, I do not use Maven anymore nor anyone else has complained about that.

**Reply by i-make-robots** [reply]
I didn't get any warnings about xml-apis.  The entire error message at run time is:
```
Error occurred during initialization of boot layer
java.lang.module.FindException: Module xalan not found, required by io.sf.carte.echosvg.dom

Process finished with exit code 1
```
I am working on branch https://github.com/MarginallyClever/Makelangelo-software/tree/replace-batik-with-css4j and running it with IJ.
All the SVG work is in `src/main/java/com/marginallyclever/makelangelo/makeart/io/vector/LoadSVG.java`

**Reply by i-make-robots** [reply]
clean and rebuild, now i get
```
Error occurred during initialization of boot layer
java.lang.module.FindException: Unable to derive module descriptor for C:\Users\aggra\.m2\repository\xalan\xalan\2.7.2\xalan-2.7.2.jar
Caused by: java.lang.module.InvalidModuleDescriptorException: Provider class org.apache.bsf.BSFManager not in module
```

**Reply by carlosame** [reply]
This output is from Maven:
```console
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ Makelangelo ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[ERROR] Error occurred during initialization of boot layer
[ERROR] java.lang.module.FindException: Module xalan not found, required by io.sf.carte.echosvg.dom
```
And then I recalled that I had this problem when I was using Maven. The module system is refusing to create a filename-based module name from the Xalan package because it is a service provider and it is not declaring a `module-info`, or something like that.

The workaround is to declare a module name with a plugin like [Moditect](https://github.com/moditect/moditect/). If you did not see a similar problem with Batik is probably due to that project not being modularised.

**Comment by carlosame** [comment]
> `Provider class org.apache.bsf.BSFManager not in module`

Exactly, the JPMS wants Xalan to have a `module-info.java` file providing the `org.apache.bsf.BSFManager`. Yet another problem caused by unmaintained ASF projects.

Perhaps I should create a EchoSVG FAQ containing this, what do you think?

**Reply by i-make-robots** [reply]
Yes.  Once we get this working a concise "how to with maven" would be great.

**Comment by i-make-robots** [comment]
https://github.com/MarginallyClever/Makelangelo-software/pull/645
We've been able to get it to run (when the moon is right) and we're still seeing the bugs of the original batik.  What do you need from us to create a real test case ?  I give you permission to use our test images in your test suite.

**Reply by carlosame** [reply]
> We've been able to get it to run (when the moon is right) and we're still seeing the bugs of the original batik. What do you need from us to create a real test case ?

I wrote that above: a valid bug report explaining what one needs to reproduce the issue with EchoSVG (not with your software).

For now, this is what I get when I render the SVG in MarginallyClever/Makelangelo-software#492 with EchoSVG:

![makelangelo](https://user-images.githubusercontent.com/38894633/162030687-c2ddc250-f148-415e-8cb8-3cabd9fe191e.png)

which seems to be the correct result.

**Comment by i-make-robots** [comment]
Well then my question is "how did you get it to run?"  I'm still trying to get around the BSFManager problem to execute the debugger in IntelliJ.

**Reply by i-make-robots** [reply]
I see that xalan-2.7.2.jar from maven central has a _META-INF/services/org.apache.xalan.extensions.bsf.BSFManager_ file declaring _org.apache.bsf.BSFManager_ as a provider ... but org.apache.bsf doesn't exist in the JAR.

**Reply by carlosame** [reply]
> I see that xalan-2.7.2.jar from maven central has a _META-INF/services/org.apache.xalan.extensions.bsf.BSFManager_ file declaring _org.apache.bsf.BSFManager_ as a provider ... but org.apache.bsf doesn't exist in the JAR.

I see your [bug](https://marc.info/?l=xalan-dev&m=164935208630930&w=2) [reports](https://issues.apache.org/jira/browse/XALANJ-2632) to the Xalan people, and raise my [pull request to Xalan-Java](https://github.com/apache/xalan-java/pull/1).

At least it has been tried.

As for how to work around this: with Gradle, either [`extra-java-module-info`](https://plugins.gradle.org/plugin/de.jjohannes.extra-java-module-info) (which is what I use) or [`moditect-gradle-plugin`](https://github.com/moditect/moditect-gradle-plugin). With Maven, [Moditect Maven plugin](https://github.com/moditect/moditect/). Weren't you already using that?

**Reply by carlosame** [reply]
And the new [FAQ](https://github.com/css4j/echosvg/blob/master/FAQ.md) talks about this...

**Comment by i-make-robots** [comment]
great.  can i please get the maven moditect settings?

On Thu, Apr 7, 2022 at 11:35 AM carlosame ***@***.***> wrote:

> And the new FAQ <https://github.com/css4j/echosvg/blob/master/FAQ.md>
> talks about this...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/css4j/echosvg/discussions/43#discussioncomment-2525936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALFRBTHZ4THXNH53O4YFGLVD4TITANCNFSM5SOPPWDQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 

Dan Royer

Owner, Marginally Clever Robots, Ltd.
------------------------------
1 (604) 259 9564 <16042599564> ***@***.***
https://marginallyclever.com <http://www.marginallyclever.com>
<https://www.facebook.com/MarginallyClever/>
<https://www.instagram.com/imakerobots/>
<https://www.youtube.com/channel/UCfbRxqjuOgE2EzRKxcePArw>
<https://twitter.com/MarginallyC> <https://github.com/MarginallyClever/>

**Comment by carlosame** [comment]
The Xalan issue is something that anyone trying to use Maven is going to find. Although my advice is always to migrate from Maven to Gradle (because Maven is not friendly to smaller, non-central repositories), the current situation is unacceptable and I decided to get rid of the Xalan dependency.

**Reply by i-make-robots** [reply]
That's awesome!  Please let me know when I can update with that change.

**Comment by i-make-robots** [comment]
```<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->
<svg width="210mm" height="297mm" viewBox="0 0 210 297" version="1.1" id="svg5" inkscape:version="1.1.1 (c3084ef, 2021-09-22)" sodipodi:docname="anniversaire.svg" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns="http://www.w3.org/2000/svg" xmlns:svg="http://www.w3.org/2000/svg">
  <sodipodi:namedview id="namedview7" pagecolor="#ffffff" bordercolor="#666666" borderopacity="1.0" inkscape:pageshadow="2" inkscape:pageopacity="0.0" inkscape:pagecheckerboard="0" inkscape:document-units="mm" showgrid="false" inkscape:zoom="0.77771465" inkscape:cx="7.7149119" inkscape:cy="576.04676" inkscape:window-width="2560" inkscape:window-height="1330" inkscape:window-x="1680" inkscape:window-y="25" inkscape:window-maximized="1" inkscape:current-layer="layer1"/>
  <defs id="defs2">
    <rect x="60.433477" y="52.718565" width="1.2858187" height="74.577482" id="rect7767"/>
  </defs>
  <g inkscape:label="Layer 1" inkscape:groupmode="layer" id="layer1">
    <g id="g888" transform="matrix(0.11652319,0,0,0.12098981,5.0275387,231.79878)">
      <path d="m 382.5,336.234 v -67.468 c 9.29,-3.138 16,-11.93 16,-22.266 0,-12.958 -10.542,-23.5 -23.5,-23.5 h -34.11 c 10.826,-10.135 17.61,-24.536 17.61,-40.5 v -80 C 358.5,71.897 333.603,47 303,47 H 278.5 V 26.886 c 4.188,-2.546 7,-7.138 7,-12.386 C 285.5,6.505 278.995,0 271,0 c -7.995,0 -14.5,6.505 -14.5,14.5 0,5.248 2.812,9.84 7,12.386 V 47 h -65 V 26.886 c 4.188,-2.546 7,-7.138 7,-12.386 C 205.5,6.505 198.995,0 191,0 c -7.995,0 -14.5,6.505 -14.5,14.5 0,5.248 2.812,9.84 7,12.386 V 47 H 159 c -30.603,0 -55.5,24.897 -55.5,55.5 v 80 c 0,15.964 6.784,30.365 17.61,40.5 H 87 c -12.958,0 -23.5,10.542 -23.5,23.5 0,10.336 6.71,19.128 16,22.266 v 67.468 c -9.29,3.138 -16,11.93 -16,22.266 v 16 c 0,4.142 3.358,7.5 7.5,7.5 4.142,0 7.5,-3.358 7.5,-7.5 v -16 c 0,-4.687 3.813,-8.5 8.5,-8.5 4.687,0 8.5,3.813 8.5,8.5 v 16 c 0,4.142 3.358,7.5 7.5,7.5 4.142,0 7.5,-3.358 7.5,-7.5 v -16 c 0,-10.336 -6.71,-19.128 -16,-22.266 V 270 h 33 v 72.5 c 0,12.958 10.542,23.5 23.5,23.5 h 9.409 c 1.245,5.075 3.72,9.669 7.093,13.449 0,0.017 -0.003,0.034 -0.003,0.051 V 415 H 159 c -21.78,0 -39.5,17.72 -39.5,39.5 0,4.142 3.358,7.5 7.5,7.5 h 88 c 4.142,0 7.5,-3.358 7.5,-7.5 v -32 c 0,-4.142 -3.358,-7.5 -7.5,-7.5 h -0.5 v -35.5 c 0,-0.017 -0.002,-0.034 -0.003,-0.051 3.374,-3.78 5.849,-8.375 7.093,-13.449 h 18.819 c 1.245,5.075 3.72,9.669 7.093,13.449 0,0.017 -0.003,0.034 -0.003,0.051 V 415 H 247 c -4.142,0 -7.5,3.358 -7.5,7.5 v 32 c 0,4.142 3.358,7.5 7.5,7.5 h 88 c 4.142,0 7.5,-3.358 7.5,-7.5 0,-21.78 -17.72,-39.5 -39.5,-39.5 h -8.5 v -35.5 c 0,-0.017 -0.002,-0.034 -0.003,-0.051 3.374,-3.78 5.849,-8.375 7.093,-13.449 H 311 c 12.958,0 23.5,-10.542 23.5,-23.5 V 270 h 33 v 66.234 c -9.29,3.138 -16,11.93 -16,22.266 v 16 c 0,4.142 3.358,7.5 7.5,7.5 4.142,0 7.5,-3.358 7.5,-7.5 v -16 c 0,-4.687 3.813,-8.5 8.5,-8.5 4.687,0 8.5,3.813 8.5,8.5 v 16 c 0,4.142 3.358,7.5 7.5,7.5 4.142,0 7.5,-3.358 7.5,-7.5 v -16 c 0,-10.336 -6.71,-19.128 -16,-22.266 z M 375,255 c -4.687,0 -8.5,-3.813 -8.5,-8.5 0,-4.687 3.813,-8.5 8.5,-8.5 4.687,0 8.5,3.813 8.5,8.5 0,4.687 -3.813,8.5 -8.5,8.5 z M 118.5,102.5 C 118.5,80.168 136.668,62 159,62 h 144 c 22.332,0 40.5,18.168 40.5,40.5 v 48 c 0,8.239 -16.171,16.553 -30.019,21.169 C 291.56,178.976 262.268,183 231,183 199.732,183 170.439,178.976 148.519,171.669 134.671,167.053 118.5,158.739 118.5,150.5 Z  

        m 0,80 v -8.518 c 5.971,4.143 14.148,8.208 25.275,11.918 23.41,7.803 54.387,12.1 87.225,12.1 32.838,0 63.815,-4.297 87.225,-12.101 11.128,-3.709 19.304,-7.775 25.275,-11.918 v 8.518 c 0,22.332 -18.168,40.5 -40.5,40.5 H 159 C 136.668,223 118.5,204.832 118.5,182.5 Z" id="path880" style=""/>
      <path d=""/>
    </g>
    <text xml:space="preserve" style="" x="20.752579" y="42.525776" id="text1743"><tspan sodipodi:role="line" id="tspan1741" style="" x="20.752579" y="42.525776">Joyeux</tspan><tspan sodipodi:role="line" style="" x="20.752579" y="75.094528" id="tspan1745">Anniversaire</tspan></text>
    <text xml:space="preserve" transform="scale(0.26458333)" id="text7765" style=""/>
  </g>
</svg>
```
loads for me like
https://cdn.discordapp.com/attachments/780846006945579058/988229462103580692/unknown.png
when it should look like
https://cdn.discordapp.com/attachments/780846006945579058/988229661127483482/unknown.png

the last part of the curve is wrong, namely 
```
m 0,80 
v -8.518 
c 5.971,4.143 14.148,8.208 25.275,11.918 23.41,7.803 54.387,12.1 87.225,12.1 32.838,0 63.815,-4.297 87.225,-12.101 11.128,-3.709 19.304,-7.775 25.275,-11.918 
v 8.518 
c 0,22.332 -18.168,40.5 -40.5,40.5 
H 159 
C 136.668,223 118.5,204.832 118.5,182.5 
Z
```
When I parse it I in my app what I receive from `SVGOMPathElement.getNormalizedPathSegList()`  is
```
Move Abs (118.5,230.5)
Line Abs (118.5,221.982)
Cubic curve (124.471,226.125)-(132.648,230.19)-(143.775,233.9)
Cubic curve (167.185,241.70299)-(198.16199,246.0)-(231.0,246.0)
Cubic curve (263.838,246.0)-(294.815,241.703)-(318.225,233.899)
Cubic curve (329.353,230.19)-(337.529,226.12401)-(343.5,221.981)
Line Abs (343.5,230.49901)
Cubic curve (343.5,252.83101)-(325.332,270.99902)-(303.0,270.99902)
Line Abs (159.0,270.99902)
Cubic curve (136.668,223.0)-(118.5,204.832)-(118.5,182.5)
Close path (71.18943363428116, 981.4940041899681, 1.0)
```
My code for this test begins here:
https://github.com/MarginallyClever/Makelangelo-software/blob/9a2c928092833100a5ca3794375202a72ad0b748/src/main/java/com/marginallyclever/makelangelo/makeart/io/vector/LoadSVG.java#L331

On the chance that I'm going blind with age, can you see what I'm doing wrong?  Did I forget a + that should be a -?

**Reply by i-make-robots** [reply]
To be clear: the previous post was made while testing with echosvg 0.2.

**Reply by i-make-robots** [reply]
What looks wrong to me is the `Move Abs` at the start, when 'm' is a relative move.  I never receive any relative moves from Batik, so I wonder if there's a mistake somewhere in there.

**Comment by i-make-robots** [comment]
Hello!  I hope you're doing well.

I had enough motivation to day to jump into this again and updated the original bug on makelangelo software with a more thorough analysis of what's going on.  I'm using EchoSVG 0.2 and getting the same glitch I had with Batik so...the bug was ported over?
I believe the issue is in `getNormalizedPathSegList`.  Where are rel moves converted to abs moves, please?

**Reply by carlosame** [reply]
> I had enough motivation to day to jump into this again and updated the original bug on makelangelo software with a more thorough analysis of what's going on. I'm using EchoSVG 0.2 and getting the same glitch I had with Batik so...the bug was ported over? I believe the issue is in `getNormalizedPathSegList`. Where are rel moves converted to abs moves, please?

Hi Dan,

I'm happy to hear that you are using EchoSVG, and it is unfortunate that you are still facing that problem.

I did not write the implementation of `getNormalizedPathSegList` and I'm not familiar with the different possible behaviors that could arise as a consequence of different configurations of the objects that are involved. My impression is that your issue is related to how some objects are configured, with certain combinations being buggy and poorly tested in EchoSVG/Batik.

As to how to fix this, IMHO debugging your issue is something that belongs to the developers in your project. My development resources are limited (I manage about 18 open source repositories and occasionally contribute to upstream dependencies), so if you could come up with a PR I could review it, or if you have a more specific suggestion I could try looking at it at some point, but that's all.

Sorry if that's disappointing but I'm not able to do a lot more, and my priorities are with the pinned issues (which are big enough already).

**Reply by i-make-robots** [reply]
alright, I respect that.  I tried digging into getNormalizedPathSegList and didn't get very far.  With the amount of code you're managing we're probably equally familiar with the underpinnings.  I've looked for Batik documentation, too, with no luck.  I'll try using `getPathSegList` (aka without normalization) and report back.

**Comment by i-make-robots** [comment]
I don't know if this qualifies as a bug yet... 
```
java.util.MissingResourceException: Unable to find resource: property.syntax.error.at
	at io.sf.carte.echosvg.i18n/io.sf.carte.echosvg.i18n.LocalizableSupport.getString(LocalizableSupport.java:328)
	at io.sf.carte.echosvg.i18n/io.sf.carte.echosvg.i18n.LocalizableSupport.formatMessage(LocalizableSupport.java:213)
	at io.sf.carte.echosvg.css/io.sf.carte.echosvg.css.engine.Messages.formatMessage(Messages.java:71)
	at io.sf.carte.echosvg.css/io.sf.carte.echosvg.css.engine.CSSEngine.getCascadedStyleMap(CSSEngine.java:731)
	at io.sf.carte.echosvg.css/io.sf.carte.echosvg.css.engine.CSSEngine.getComputedStyle(CSSEngine.java:819)
	at io.sf.carte.echosvg.bridge/io.sf.carte.echosvg.bridge.CSSUtilities.getComputedStyle(CSSUtilities.java:85)
	at io.sf.carte.echosvg.bridge/io.sf.carte.echosvg.bridge.CSSUtilities.convertDisplay(CSSUtilities.java:523)
	at io.sf.carte.echosvg.bridge/io.sf.carte.echosvg.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:204)
	at io.sf.carte.echosvg.bridge/io.sf.carte.echosvg.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at io.sf.carte.echosvg.bridge/io.sf.carte.echosvg.bridge.GVTBuilder.build(GVTBuilder.java:84)
	at com.marginallyclever.makelangelo/com.marginallyclever.makelangelo.makeart.io.vector.LoadSVG.initSVGDOM(LoadSVG.java:647)
```
when trying to load
```xml
<svg xmlns="http://www.w3.org/2000/svg" width="320" height="320">
  <path d="M 10 315
           L 110 215
           A 36 60 0 0 1 150.71 170.29
           L 172.55 152.45
           A 30 50 -45 0 1 215.1 109.9
           L 315 10" stroke="black"/>
  <circle cx="150.71" cy="170.29" r="2" fill="red"/>
  <circle cx="110" cy="215" r="2" fill="red"/>
  <ellipse cx="144.931" cy="229.512" rx="36" ry="60" fill="transparent" stroke="blue"/>
  <ellipse cx="115.779" cy="155.778" rx="36" ry="60" fill="transparent" stroke="blue"/>
</svg>
```
ellipse elements seem to make it mad.

**Reply by carlosame** [reply]
Looks like you are having a classloader problem (the MissingResourceException). What I get is the following:

```java
The attribute "fill" represents an invalid CSS value ("transparent").
Original message:
The "transparent" identifier is not a valid value for the "fill" property. 
org.w3c.dom.DOMException: file:/C:/***/echosvg/samples/ellipses.svg:
The attribute "fill" represents an invalid CSS value ("transparent").
Original message:
The "transparent" identifier is not a valid value for the "fill" property. 
	at io.sf.carte.echosvg.css.engine.CSSEngine.getCascadedStyleMap(CSSEngine.java:733)
	at io.sf.carte.echosvg.css.engine.CSSEngine.getComputedStyle(CSSEngine.java:819)
	at io.sf.carte.echosvg.bridge.CSSUtilities.getComputedStyle(CSSUtilities.java:85)
	at io.sf.carte.echosvg.bridge.CSSUtilities.convertDisplay(CSSUtilities.java:523)
	at io.sf.carte.echosvg.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:204)
	at io.sf.carte.echosvg.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171)
	at io.sf.carte.echosvg.bridge.GVTBuilder.build(GVTBuilder.java:84)
	at io.sf.carte.echosvg.transcoder.SVGAbstractTranscoder.transcode(SVGAbstractTranscoder.java:241)
	at io.sf.carte.echosvg.transcoder.image.ImageTranscoder.transcode(ImageTranscoder.java:97)
```
So the problem is caused by the lack of support for the `transparent` identifier. I'll see what I can do.

**Reply by carlosame** [reply]
EchoSVG now supports the `transparent` color identifier and can render your SVG file just fine.

**Comment by carlosame** [comment]
Given that this discussion is receiving some attention these days, I wanted to point out that the issue with `getNormalizedPathSegList` was fixed in a7adb2a2341e7a5197be8bd797f576d87d6296cf (see #50). I am closing this discussion as some people may believe that the issue still exists.
